File: backend/app/schemas/documents.py
"""Document schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class DocumentResponse(BaseModel):
    """Document response payload."""

    id: UUID
    owner_id: UUID
    workspace_id: UUID | None = None
    title: str
    authors: list[str] = Field(default_factory=list)
    year: int | None = None
    doi: str | None = None
    file_path: str
    file_size: int
    page_count: int | None = None
    status: str
    processing_progress: int = 0
    created_at: datetime
    
    # Document Intelligence Metrics
    text_coverage_pct: float | None = None
    ocr_coverage_pct: float | None = None
    chunk_count: int = 0
    embedding_count: int = 0
    searchable_status: str = "Pending"


class DocumentListResponse(BaseModel):
    """List of documents."""

    items: list[DocumentResponse]


class DocumentUploadResponse(BaseModel):
    """Document upload response."""

    document: DocumentResponse


class DocumentUpdate(BaseModel):
    """Document update payload."""
    title: str | None = None
    workspace_id: UUID | None = None
    authors: list[str] | None = None
    year: int | None = None
    doi: str | None = None

File: frontend/src/lib/api/documents.ts
import { apiClient } from "./client";

export interface DocumentUploadResponse {
  id: string;
  title: string;
  filename: string;
  status: string;
}

export interface DocumentItem {
  id: string;
  title: string;
  filename: string;
  authors: string[];
  workspace_id?: string | null;
  year?: number;
  doi?: string;
  status: string;
  processing_progress: number;
  created_at: string;
  text_coverage_pct?: number | null;
  ocr_coverage_pct?: number | null;
  chunk_count: number;
  embedding_count: number;
  searchable_status: string;
  page_count: number;
  file_size?: number;
}

export interface DocumentListResponse {
  items: DocumentItem[];
  total: number;
}

export const documentsApi = {
  upload: (data: FormData) => {
    const authData = JSON.parse(localStorage.getItem("auth-storage") || "{}");
    const token = authData?.state?.accessToken;
    const headers = new Headers();
    if (token) {
      headers.set("Authorization", `Bearer ${token}`);
    }
    
    // We don't use apiClient here because fetch handles multipart/form-data 
    // boundary automatically when Content-Type is NOT set.
    return fetch(`${process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"}/documents/upload`, {
      method: "POST",
      headers,
      body: data
    }).then(res => {
      if (!res.ok) throw new Error("Upload failed");
      return res.json() as Promise<DocumentUploadResponse>;
    });
  },
  
  list: () => apiClient<DocumentListResponse>("/documents", {
    method: "GET"
  }),

  getDocument: (id: string) => apiClient<DocumentItem>(`/documents/${id}`, {
    method: "GET"
  }),

  delete: (id: string) => apiClient<{ status: string }>(`/documents/${id}`, {
    method: "DELETE"
  }),

  update: (id: string, data: { title?: string; authors?: string[]; year?: number; doi?: string; workspace_id?: string | null }) => 
    apiClient<DocumentItem>(`/documents/${id}`, {
      method: "PATCH",
      body: JSON.stringify(data)
    }),

  getContentBlob: async (id: string) => {
    const authData = JSON.parse(localStorage.getItem("auth-storage") || "{}");
    const token = authData?.state?.accessToken;
    const headers = new Headers();
    if (token) {
      headers.set("Authorization", `Bearer ${token}`);
    }
    
    const res = await fetch(`${process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"}/documents/${id}/content`, {
      method: "GET",
      headers
    });
    
    if (!res.ok) throw new Error("Failed to download document");
    return res.blob();
  }
};
